hoomd/hpmc/pytest/test_faceted_ellipsoid.py

```diff
@@ -1,31 +1,314 @@
-    test_faceted_ellipsoid5 = hpmc.faceted_ellipsoid_params(args_5)
+    test_faceted_ellipsoid5 = hpmc.FacetedEllipsoidParams(args_5)
     test_dict5 = test_faceted_ellipsoid5.asDict()
     assert test_dict5 == args_5
 
+
+def test_shape_params():
+
+    mc = hoomd.hpmc.integrate.FacetedEllipsoid(23456)
+
+    mc.shape['A'] = dict()
+    assert mc.shape['A']['normals'] == hoomd.typeconverter.RequiredArg
+    assert mc.shape['A']['offsets'] == hoomd.typeconverter.RequiredArg
+    assert mc.shape['A']['vertices'] == hoomd.typeconverter.RequiredArg
+    assert mc.shape['A']['origin'] == hoomd.typeconverter.RequiredArg
+    assert mc.shape['A']['a'] == hoomd.typeconverter.RequiredArg
+    assert mc.shape['A']['b'] == hoomd.typeconverter.RequiredArg
+    assert mc.shape['A']['c'] == hoomd.typeconverter.RequiredArg
+    assert mc.shape['A']['ignore_statistics'] is False
+
+    mc.shape['B'] = dict(a=2.5, b=1, c=3)
+    assert mc.shape['B']['normals'] == hoomd.typeconverter.RequiredArg
+    assert mc.shape['B']['offsets'] == hoomd.typeconverter.RequiredArg
+    assert mc.shape['B']['vertices'] == hoomd.typeconverter.RequiredArg
+    assert mc.shape['B']['origin'] == hoomd.typeconverter.RequiredArg
+    assert mc.shape['B']['a'] == 2.5
+    assert mc.shape['B']['b'] == 1
+    assert mc.shape['B']['c'] == 3
+    assert mc.shape['B']['ignore_statistics'] is False
+
+    mc.shape['C'] = args_1
+    for key in args_1.keys():
+        assert mc.shape['C'][key] == args_1[key]
+
+    mc.shape['D'] = args_2
+    for key in args_2.keys():
+        assert mc.shape['D'][key] == args_2[key]
+
+    mc.shape['E'] = args_3
+    for key in args_3.keys():
+        assert mc.shape['E'][key] == args_3[key]
+
+    mc.shape['F'] = args_4
+    for key in args_4.keys():
+        assert mc.shape['F'][key] == args_4[key]
+
+    mc.shape['G'] = args_5
+    for key in args_5.keys():
+        assert mc.shape['G'][key] == args_5[key]
+
+
+def test_shape_params_attached(device, dummy_simulation_factory):
+
+    mc = hoomd.hpmc.integrate.FacetedEllipsoid(23456)
+
+    mc.shape['A'] = args_1
+    mc.shape['B'] = args_2
+    mc.shape['C'] = args_3
+    mc.shape['D'] = args_4
+    mc.shape['E'] = args_5
+
+    sim = dummy_simulation_factory(particle_types=['A', 'B', 'C', 'D', 'E'])
+    sim.operations.add(mc)
+    sim.operations.schedule()
+
+    assert mc.shape['A']['ignore_statistics']
+    assert not mc.shape['B']['ignore_statistics']
+    for key in args_1.keys():
+        assert mc.shape['A'][key] == args_1[key]
+        assert mc.shape['B'][key] == args_2[key]
+        assert mc.shape['C'][key] == args_3[key]
+        assert mc.shape['D'][key] == args_4[key]
+        assert mc.shape['E'][key] == args_5[key]
+
+    args_1_invalid = copy.deepcopy(args_1)
```

Comment:
If using pytest fixtures with a function scope (the default), you wouldn't need to work about deepcopying since they are regenerated after every test.